hail/src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala

```diff
@@ -82,54 +82,71 @@ object IntervalFunctions extends RegistryFunctions {
+      case (r, rt, interval: PIntervalCode) =>
+        val empty = EmitCodeBuilder.scopedCode(r.mb) { cb =>
+          val intv = interval.memoize(cb, "interval")
+          intv.isEmpty(cb)
+        }
+        PCode(rt, empty)
     }
 
-    registerCode1("isEmpty", TInterval(tv("T")), TBoolean, (_: Type, pt: PType) => PBoolean(pt.required)) {
-      case (r, rt, (intervalT: PInterval, intOff)) =>
-        val interval = new IRInterval(r, intervalT, intOff)
+    registerPCode2("overlaps", TInterval(tv("T")), TInterval(tv("T")), TBoolean, (_: Type, i1t: PType, i2t: PType) => PBoolean(i1t.required && i2t.required)) {
+      case (r, rt, int1: PIntervalCode, int2: PIntervalCode) =>
+        val overlap = EmitCodeBuilder.scopedCode(r.mb) { cb =>
+          val interval1 = int1.memoize(cb, "interval1")
+          val interval2 = int2.memoize(cb, "interval2")
+          val compare = cb.emb.getCodeOrdering(int1.pt.pointType, int2.pt.pointType, CodeOrdering.Compare())
+
+          def isAboveOnNonempty(cb: EmitCodeBuilder, lhs: PIntervalValue, rhs: PIntervalValue): Code[Boolean] = {
+            val cmp = r.mb.newLocal[Int]()
+            val start = EmitCode.fromI(cb.emb)(lhs.loadStart(_))
+            val end = EmitCode.fromI(cb.emb)(rhs.loadEnd(_))
+            cb += start.setup
+            cb += end.setup
+            cb.assign(cmp, compare(start.m -> start.v, end.m -> end.v))
+            cmp > 0 || (cmp.ceq(0) && (!lhs.includesStart() || !rhs.includesEnd()))
+          }
 
-        Code(
-          interval.storeToLocal,
-          interval.isEmpty
-        )
-    }
+          def isBelowOnNonempty(cb: EmitCodeBuilder, lhs: PIntervalValue, rhs: PIntervalValue): Code[Boolean] = {
+            val start = EmitCode.fromI(cb.emb)(rhs.loadStart(_))
+            val end = EmitCode.fromI(cb.emb)(lhs.loadEnd(_))
+            cb += start.setup
+            cb += end.setup
+            val cmp = r.mb.newLocal[Int]()
+            cb.newLocal("cmp", compare(start.m -> start.v, end.m -> end.v))
+            cmp < 0 || (cmp.ceq(0) && (!lhs.includesEnd() || !rhs.includesStart()))
+          }
 
-    registerCode2("overlaps", TInterval(tv("T")), TInterval(tv("T")), TBoolean, (_: Type, i1t: PType, i2t: PType) => PBoolean(i1t.required && i2t.required)) {
-      case (r, rt, (i1t: PInterval, iOff1), (i2t: PInterval, iOff2)) =>
-        val interval1 = new IRInterval(r, i1t, iOff1)
-        val interval2 = new IRInterval(r, i2t, iOff2)
-
-        Code(
-          interval1.storeToLocal,
-          interval2.storeToLocal,
-          !(interval1.isEmpty || interval2.isEmpty ||
-            interval1.isBelowOnNonempty(interval2) ||
-            interval1.isAboveOnNonempty(interval2))
-        )
+          interval1.isEmpty(cb) || interval2.isEmpty(cb) ||
+            isBelowOnNonempty(cb, interval1, interval2) ||
+            isAboveOnNonempty(cb, interval1, interval2)
```

Comment:
I think this is going to result in generating the setup code for the start and end of both intervals twice. I'm not sure if that's a bad thing, but wanted to flag it and ask.

Reply:
it's memoized on line 121, this should be fine I think

Reply:
So I see that it's memoized, but then it's converted into an `EmitCode` and the `setup` is added to the code builder in both methods. Is the expectation that the `setup`s are empty (or at least tiny) for something that is coming from a memoized value?

Reply:
1. EmitCode.fromI currently always produces an empty setup.
2. This should be no worse than the old code in that it also does a double load, once for the `isEmpty`, and again for the `isAbove/isBelow` calls. An extra memoize to turn these loads into variable loads could be done, but I feel it's largely unnecessary.

Reply:
Just wanted to make sure I understood. Reasoning makes sense